company-files.el

```diff
@@ -61,30 +63,36 @@
   (or (not (file-remote-p file))
       (file-remote-p file nil t)))
 
+(defun company-files--trailing-slash-p (file)
+  ;; `file-directory-p' is very expensive on remotes. We are relying on
+  ;; `file-name-all-completions' returning directories with trailing / instead.
+  (let ((len (length file)))
+    (and (> len 0) (eq (aref file (1- len)) ?/))))
+
 (defvar company-files--completion-cache nil)
 
 (defun company-files--complete (prefix)
   (let* ((dir (file-name-directory prefix))
-         (key (list (file-name-nondirectory prefix)
+         (file (file-name-nondirectory prefix))
+         (key (list file
                     (expand-file-name dir)
                     (nth 5 (file-attributes dir))))
-         (file (file-name-nondirectory prefix))
-         (completion-ignore-case read-file-name-completion-ignore-case)
-         candidates directories)
+         (completion-ignore-case read-file-name-completion-ignore-case))
     (unless (company-file--keys-match-p key (car company-files--completion-cache))
-      (dolist (file (company-files--directory-files dir file))
-        (setq file (concat dir file))
-        (when (company-files--connected-p file)
-          (push file candidates)
-          (when (file-directory-p file)
-            (push file directories))))
-      (dolist (directory (reverse directories))
-        ;; Add one level of children.
-        (dolist (child (company-files--directory-files directory ""))
-          (push (concat directory
-                        (unless (eq (aref directory (1- (length directory))) ?/) "/")
-                        child) candidates)))
-      (setq company-files--completion-cache (cons key (nreverse candidates))))
+      (let* ((candidates (mapcar (lambda (f) (concat dir f))
+                                 (company-files--directory-files dir file)))
+             (directories (cl-remove-if-not (lambda (f)
+                                              (and (company-files--trailing-slash-p f)
+                                                   (company-files--connected-p f)))
+                                            candidates))
+             (children (unless (file-remote-p dir)
+                         (cl-mapcan (lambda (d)
+                                      (mapcar (lambda (c) (concat d c))
+                                              (company-files--directory-files d "")))
+                                    directories))))
+        (setq company-files--completion-cache
+              (cons key (sort (append candidates children)
```

Comment:
For two reasons. First, `file-name-all-completions` doesn't return a sorted list. Second, for readability. Sideeffecting in dolist is difficult to reason about.

Sort is pretty cheap as compared to other stuff. It's about 1ms for 5000 files on my machine. So I would not be too concerned about it anyways.